### PR TITLE
fix #13297 fix crashes on reset preferences

### LINKED ISSUE #13297: [MU4 Issue] Reset Preferences crashes app if a part with no instruments is open
**Describe the bug**
If you've created an "empty" part with no instruments, reset preferences causes the app to crash.

**To Reproduce**
1. Open/create any score
2. From the Parts dialog, click Create New, then open selected.
3. Go to Edit|Preferences and select "Reset preferences"
4. App crashes

**Expected behavior**
App does not crash

**Screenshots**

**Platform information**
Windows 10

**Additional context**
Built from master branch, 12-Sep-2022.
Stack trace:

Exception thrown: read access violation
*this* was nullptr

>	MuseScore4.exe!mu::engraving::Score::doLayoutRange(const mu::engraving::Fraction & st, const mu::engraving::Fraction & et) Line 5500	C++
 	MuseScore4.exe!mu::engraving::Score::doLayout() Line 5493	C++
 	MuseScore4.exe!mu::notation::Notation::{ctor}::__l2::<lambda>(mu::framework::Orientation __formal) Line 95	C++
 	[External Code]	
 	MuseScore4.exe!deto::async::Channel<enum mu::framework::Orientation>::ReceiveCall<void <lambda>(mu::framework::Orientation),enum mu::framework::Orientation>::received(const deto::async::NotifyData & d) Line 80	C++
 	MuseScore4.exe!deto::async::Channel<enum mu::framework::Orientation>::ChannelInvoker::doInvoke(int callKey, void * call, const deto::async::NotifyData & d) Line 127	C++
 	MuseScore4.exe!deto::async::AbstractInvoker::invokeCallback(int type, const deto::async::AbstractInvoker::CallBack & c, const deto::async::NotifyData & data) Line 62	C++
 	MuseScore4.exe!deto::async::AbstractInvoker::invoke(int type, const deto::async::NotifyData & data) Line 45	C++
 	MuseScore4.exe!deto::async::Channel<enum mu::framework::Orientation>::send(const mu::framework::Orientation & <d_0>) Line 32	C++

This is because m_score is actually nullptr for the Notation object which sets up a "canvasOrientation" receiver (so this would likely happen when changing canvas orientation, and potentially other operations). I'm not sure if we should ever have a Notation object with m_score being nullptr, but obviously it is happening so we should handle it.

### PR DESCRIPTION
**Pull request by wizofaus**
Resolves: https://github.com/musescore/MuseScore/issues/13297

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: src/appshell/view/preferences/preferencesmodel.cpp [hunk 1]
```diff
@@ -198,8 +199,11 @@ void PreferencesModel::load(const QString& currentPageId)
 void PreferencesModel::resetFactorySettings()
 {
     static constexpr bool KEEP_DEFAULT_SETTINGS = true;
+    QApplication::setOverrideCursor(Qt::WaitCursor);
+    QApplication::processEvents();
     configuration()->revertToFactorySettings(KEEP_DEFAULT_SETTINGS);
     configuration()->startEditSettings();
+    QApplication::restoreOverrideCursor();
```
Comment (wizofaus) [inline src/appshell/view/preferences/preferencesmodel.cpp:206]: not strictly relevant to the crash but it always felt like it was hanging without this

File: src/engraving/libmscore/select.cpp [hunk 2]
```diff
@@ -531,7 +531,7 @@ void Selection::updateSelectedElements()
         LOGE() << "selection locked, reason: " << lockReason();
         return;
     }
-    if (_state != SelState::RANGE) {
+    if (_state != SelState::RANGE || !startSegment()) {
```
Comment (wizofaus) [inline src/engraving/libmscore/select.cpp:534]: Also fixes https://github.com/musescore/MuseScore/issues/13291

**Comment by wizofaus** [comment]
The utests failure is nothing to do with this PR, not sure how to trigger it to re-run?

**Comment by HemantAntony** [comment]
I think rebasing it should work

**Comment by wizofaus** [comment]
Thought I had branched this off the latest master, but apparently not, so yeah, did a rebase.

**Comment by wizofaus** [comment]
Now it seems to be randomly failing on a different test, though I can't even figure out which one (in iex_musicxml_tests apparently)

**Comment by wizofaus** [comment]
I honestly have no idea what's going on here, that's 3 times now the utests have failed with this branch, so something on it must be triggering that but I can't say see how anything here would explain that, or determine which test is actual failing, or reproduce running locally.

File: src/engraving/libmscore/select.cpp [hunk 3]
```diff
@@ -624,7 +624,7 @@ void Selection::updateSelectedElements()
             }
         }
     }
-    Fraction stick = startSegment()->tick();
+    Fraction stick = tickStart();
```
Comment (wizofaus) [inline src/engraving/libmscore/select.cpp:627]: Also fixes crash mentioned in https://github.com/musescore/MuseScore/issues/13291

**Comment by RomanPudashkin** [comment]
Please rebase your PR

File: src/notation/internal/notation.cpp [hunk 4]
```diff
@@ -91,9 +91,11 @@ Notation::Notation(mu::engraving::Score* score)
     });
 
     configuration()->canvasOrientation().ch.onReceive(this, [this](framework::Orientation) {
```
[the remark this answers is not in the document]
Comment (RomanPudashkin) [reply src/notation/internal/notation.cpp:93]: Or we can keep this code in Notation::Notation(), but rewrite it as follows:

```
configuration()->canvasOrientation().ch.onReceive(this, [this](framework::Orientation) {
        if (m_score) {
            m_score->doLayout();
        }
});
```

[the remark this answers is not in the document]
Comment (wizofaus) [reply src/notation/internal/notation.cpp:93]: Sure but all I was trying to do is prevent the crash with minimal change in other behaviour.

[the remark this answers is not in the document]
Comment (RomanPudashkin) [reply src/notation/internal/notation.cpp:93]: I believe it is not needed here. It was added at the beginning of the 4.0 development, but now it looks redundant

[the remark this answers is not in the document]
Comment (wizofaus) [reply src/notation/internal/notation.cpp:93]: Well that can be fixed separately - the null check is definitely needed. I'm not able to test whether removing that loop is needed anymore currently.

**Comment by DmitryArefiev** [comment]
Tested #13297 on Win10, Mac12, LinuxMint 20.2 - **FIXED**

**Pull request merged**